lib/stdlib/src/argparse.erl

```diff
@@ -0,0 +1,1305 @@
+        Atom
+    catch error:badarg ->
+        throw({Eos#eos.commands, Opt, Arg, <<"is not an existing atom">>})
+    end;
+convert_type({custom, Fun}, Arg, Opt, Eos) ->
+    try Fun(Arg)
+    catch error:badarg ->
+        throw({Eos#eos.commands, Opt, Arg, <<"failed faildation">>})
+    end.
+
+%% Given Var, and list of {min, X}, {max, Y}, ensure that
+%%  value falls within defined limits.
+minimax(Var, [], _Eos, _Opt, _Orig) ->
+    Var;
+minimax(Var, [{min, Min} | _], Eos, Opt, Orig) when Var < Min ->
+    throw({Eos#eos.commands, Opt, Orig, <<"is less than accepted minimum">>});
+minimax(Var, [{max, Max} | _], Eos, Opt, Orig) when Var > Max ->
+    throw({Eos#eos.commands, Opt, Orig, <<"is greater than accepted maximum">>});
+minimax(Var, [Num | Tail], Eos, Opt, Orig) when is_number(Num) ->
+    lists:member(Var, [Num|Tail]) orelse
+        throw({Eos#eos.commands, Opt, Orig, <<"is not one of the choices">>}),
+    Var;
+minimax(Var, [_ | Tail], Eos, Opt, Orig) ->
+    minimax(Var, Tail, Eos, Opt, Orig).
+
+%% returns integer from string, or errors out with debugging info
+get_int(Arg, Opt, Eos) ->
+    case string:to_integer(Arg) of
+        {Int, []} ->
+            Int;
+        _ ->
+            throw({Eos#eos.commands, Opt, Arg, <<"is not an integer">>})
+    end.
+
+%% returns float from string, that is floating-point, or integer
+get_float(Arg, Opt, Eos) ->
+    case string:to_float(Arg) of
+        {Float, []} ->
+            Float;
+        _ ->
+            %% possibly in disguise
+            case string:to_integer(Arg) of
+                {Int, []} ->
+                    Int;
+                _ ->
+                    throw({Eos#eos.commands, Opt, Arg, <<"is not a number">>})
+            end
+    end.
+
+%% Returns 'true' if String can be converted to a number
+is_digits(String) ->
+    case string:to_integer(String) of
+        {_Int, []} ->
+            true;
+        {_, _} ->
+            case string:to_float(String) of
+                {_Float, []} ->
+                    true;
+                {_, _} ->
+                    false
+            end
+    end.
+
+%% 'maybe' nargs for an option that does not have default set still have
+%%  to produce something, let's call it hardcoded default.
+default(#{default := Default}) ->
+    Default;
+default(#{type := boolean}) ->
+    true;
+default(#{type := integer}) ->
+    0;
+default(#{type := float}) ->
+    0.0;
+default(#{type := string}) ->
+    "";
+default(#{type := binary}) ->
+    <<"">>;
+default(#{type := atom}) ->
+    undefined;
+%% no type given, consider it 'undefined' atom
+default(_) ->
+    undefined.
+
+%% command path is now in direct order
+format_path(Commands) ->
+    lists:join(" ", Commands).
+
+%%--------------------------------------------------------------------
+%% Validation and preprocessing
+%% Theoretically, Dialyzer should do that too.
```

Comment:
Am I right in thinking the below "just" generates clear error messages, rather than actually performs some static analysis? If so, perhaps this comment is more confusing than anything else.

Also, it might be nice to actually export the validation functions, so they can be run at build time or in-IDE by users?